Software/Program/Code/Core/Src/main.c

```diff
@@ -90,18 +147,121 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_TIM2_Init();
+  MX_ADC3_Init();
   /* USER CODE BEGIN 2 */
 
+   HAL_TIM_Base_Start_IT(&htim2);
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
-  while (1)
-  {
+   while (1)
+   {
+      static enum Direction_E Direction = DIRECTION_NOT_AVAILABLE;
+      static bool UpdateOutput = false;
+      static bool ADC_ConversionInProgress = false;
+      static uint16_t Previous_ADCValue = 0u;
+      static float Frequency = 0.0f;
+
+      /*** HANDLE FRQ_OUT GPIO ***/
```

Comment:
```suggestion
      /*** Handle Frequency Output GPIO ***/
```
Don't caps-lock the section header comments. Leave that style to the Cube.